File: api/get-user-status.py
import json
import jwt
import logging

from api.utils import get_db_connection, release_db_connection, get_cors_headers, cors_preflight, JWT_SECRET

def handler(request):
    """
    Retrieves the user's subscription status.
    """
    preflight_response = cors_preflight(request)
    if preflight_response:
        return preflight_response

    headers = get_cors_headers()
    conn = None
    try:
        auth_header = request.headers.get('Authorization')
        if not auth_header or not auth_header.startswith('Bearer '):
            return (json.dumps({'error': 'Authorization token is missing or invalid.'}), 401, headers)
            
        token = auth_header.split(' ')[1]
        payload = jwt.decode(token, JWT_SECRET, algorithms=['HS256'])
        username = payload.get('username')
        
        if not username:
            return (json.dumps({'error': 'Invalid token payload.'}), 401, headers)
            
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute('SELECT username, subscription_expires_at FROM users WHERE username = %s', (username,))
        user_record = cursor.fetchone()
        
        if not user_record:
            return (json.dumps({'error': 'User not found.'}), 404, headers)
            
        user_data = {
            'username': user_record[0],
            'subscription_expires_at': user_record[1].isoformat() if user_record[1] else None
        }
        
        return (json.dumps(user_data), 200, headers)
        
    except (jwt.ExpiredSignatureError, jwt.InvalidTokenError):
        return (json.dumps({'error': 'Token is expired or invalid.'}), 401, headers)
    except Exception as e:
        logging.error(f"Get user status error: {e}")
        return (json.dumps({'error': 'Internal server error'}), 500, headers)
    finally:
        if conn:
            try:
                cursor.close()
            except Exception as e:
                logging.error(f"Error closing cursor: {e}")
            release_db_connection(conn)

File: api/get-user-status.js
import { db } from '/lib/db';
import jwt from 'jsonwebtoken'; // 我们需要jsonwebtoken库来验证token

export default async function handler(req, res) {
    if (req.method !== 'GET') {
        return res.status(405).json({ error: 'Method Not Allowed' });
    }

    try {
        // 1. 从请求头中获取 Authorization token
        const authHeader = req.headers.authorization;
        if (!authHeader || !authHeader.startsWith('Bearer ')) {
            return res.status(401).json({ error: 'Authorization token is missing or invalid.' });
        }

        const token = authHeader.split(' ')[1];
        const jwtSecret = process.env.JWT_SECRET;

        // 2. 验证并解码 token
        let decoded;
        try {
            decoded = jwt.verify(token, jwtSecret);
        } catch (error) {
            // 如果token过期或无效，返回401 Unauthorized
            return res.status(401).json({ error: 'Token is expired or invalid.' });
        }
        
        const username = decoded.username;
        if (!username) {
            return res.status(401).json({ error: 'Invalid token payload.' });
        }

        // 3. 从数据库中查询用户信息
        const result = await db.query(
            'SELECT username, subscription_expires_at FROM users WHERE username = $1',
            [username]
        );

        if (result.rows.length === 0) {
            return res.status(404).json({ error: 'User not found.' });
        }

        const user = result.rows[0];

        // 4. 将安全的用户信息返回给前端
        res.status(200).json({
            username: user.username,
            subscription_expires_at: user.subscription_expires_at,
        });

    } catch (error) {
        console.error('Get User Status API Error:', error);
        res.status(500).json({ error: 'An internal server error occurred.' });
    }

}
